db/schema.rb

```diff
@@ -0,0 +1,22 @@
+# This file is auto-generated from the current state of the database. Instead
+# of editing this file, please use the migrations feature of Active Record to
+# incrementally modify your database, and then regenerate this schema definition.
+#
+# This file is the source Rails uses to define your schema when running `bin/rails
+# db:schema:load`. When creating a new database, `bin/rails db:schema:load` tends to
+# be faster and is potentially less error prone than running all of your
+# migrations from scratch. Old migrations may fail to apply correctly if those
+# migrations use external dependencies or application code.
+#
+# It's strongly recommended that you check this file into your version control system.
+
+ActiveRecord::Schema.define(version: 2021_11_14_005104) do
+
+  create_table "users", charset: "utf8mb4", force: :cascade do |t|
+    t.string "name"
+    t.string "email"
+    t.datetime "created_at", precision: 6, null: false
+    t.datetime "updated_at", precision: 6, null: false
+  end
+
+end
```

Comment:
- Tìm hiểu về file migration, và cách sử dụng? (rollback, create, reset)

Reply:
Ý nghĩa của các số trong file migration

Reply:
1. Migration
Migration là một tính năng của Active Record cho phép thay đổi cả cấu trúc và dữ liệu trong database một cách dễ dàng thông qua các Ruby DSL mà không cần vào trực tiếp database.
Ban đầu schema là rỗng và mỗi lần migration thì sẽ modify để add hoặc remove table, columns hoặc rows. Active record cũng sẽ update file db/schema.rb để làm thống nhất với cấu trúc mới nhất của database. Run migration thì table sẽ được tự động sinh ra. Có thể back migration trước bằng cách sử dụng rollback.
- timestamps sẽ thêm vào table 2 column created_at và update_at.

2. Create Migration
####Create migration
- rails generate migration <tên migration>
TH tên migration có định dạng "AddXXXToYYY" & "RemoveXXXFromYYY".  sau đó nếu nhập tiếp tên column thì migration sẽ tự động add thêm các trường add_colum và remove_column trong class.
- rails generate migration <tên migration> <tên column> : <kiểu dữ liệu>

####Tạo model
- rails generate model <tên model> <tên column>:<kiểu dữ liệu>
####Run Active record
- rails db:migrate
####Rollback
Nhiều trường hợp muốn Rollback lại version cũ: 
- rails db:rollback
Muốn rollback lại nhiều hơn các version
- rails db:rollback STEP=3
####create database, đọc schema và thực hiện các thiết lập đầu tiền của db
- rails db:setup
####Reset database=> sẽ drop và setting lại db
- rails db:reset <=> rails db:drop and rails db:reset 
Ngoài ra có thể sử dụng phương thức up down: Phương thức up sẽ được thực hiện khi migration chạy, trong khi down sẽ được chạy khi rollback. Có thể dùng change để viết tương  tự như các hoạt động của up down này

### Ý nghĩa các con số trong file
2021_11_14_005104: YYYY_MM_DD_HHMMSS
precision: 6: độ chính xác đến 6 chữ số
Tài liệu tham khảo:
https://viblo.asia/p/migration-trong-rails-Do754WgLlM6d
https://viblo.asia/p/hieu-sau-hon-ve-rails-migration-Eb85onrWl2G